Image upload fails with elementX

# Issue

Hi, uploading images with elementX does not work, while it does with element, so I do not believe this is a server issue. Furthermore the logs report a `Request must specify a Content-Length` error, which seems a client side misconfiguration to me

I'm using caddy as my reverse proxy, with a fix for TLS1.2 clients due to https://github.com/vector-im/element-x-ios/issues/786

```
chat.fuo.fi {

import bind-ipv6-public
import bind-ipv4-public

reverse_proxy /_matrix/* http://localmatrix:8008
reverse_proxy /_synapse/client/* http://localmatrix:8008

request_body {
  max_size 100MB
}


import fix-elementx

# HSTS (63072000 seconds)
header / Strict-Transport-Security "max-age=63072000"
}
```

```
(fix-elementx) {

@tls12only {
header_regexp User-Agent Element X
}

tls @tls12only {
protocols tls1.2 tls1.2
}


}
```

# Logs

```
2023-08-08T16:08:23.084759Z  INFO root: elementx: MediaUploadingPreprocessor.swift:259: No GPS metadata found. Returning original image
2023-08-08 18:08:23.943248+0200 ElementX[13991:2406456] [SwiftUI] Publishing changes from background threads is not allowed; make sure to publish values from the main thread (via operators like receive(on:)) on model updates.
2023-08-08T16:08:23.944621Z DEBUG root:send_attachment:send{homeserver="https://chat.fuo.fi/" server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6] config=RequestConfig { timeout: 300s } request_id="REQ-92" request_size="71.3 kiB" path="/_matrix/media/v3/upload"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:198: Sending request
2023-08-08T16:08:23.970961Z DEBUG root:send_attachment:send{homeserver="https://chat.fuo.fi/" server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6] config=RequestConfig { timeout: 300s } request_id="REQ-93" request_size="2.4 MiB" path="/_matrix/media/v3/upload"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:198: Sending request
2023-08-08 18:08:23.982618+0200 ElementX[13991:2403520] [default] Image analysis failed with error Error Domain=com.apple.VisionKit.ImageAnalyzer Code=-4 "Device Not Supported" UserInfo={NSLocalizedDescription=Device Not Supported} #ImageAnalysis
2023-08-08T16:08:24.100427Z DEBUG root:send_attachment:send{homeserver="https://chat.fuo.fi/" server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6] config=RequestConfig { timeout: 300s } request_id="REQ-92" request_size="71.3 kiB" path="/_matrix/media/v3/upload" status=400 response_size="71 B"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:208: Error while sending request: Api(Server(ClientApi(Error { status_code: 400, body: Standard { kind: Unknown, message: "Request must specify a Content-Length" } })))
2023-08-08T16:08:24.101315Z ERROR root: elementx: MediaUploadPreviewScreenViewModel.swift:63: Failed sending attachment with error: failedSendingMedia
2023-08-08 18:08:24.137048+0200 ElementX[13991:2406456] [default] Image analysis failed with error Error Domain=com.apple.VisionKit.ImageAnalyzer Code=-4 "Device Not Supported" UserInfo={NSLocalizedDescription=Device Not Supported} #ImageAnalysis

```

# Screenshot
<img width="365" alt="image" src="https://github.com/matrix-org/matrix-rust-sdk/assets/1580624/12020ff0-bc42-48c9-b17f-85aa5141cebd">

--- a/crates/matrix-sdk/src/http_client/native.rs
+++ b/crates/matrix-sdk/src/http_client/native.rs
@@ -23,6 +23,7 @@ use backoff::{future::retry, Error as RetryError, ExponentialBackoff};
 use bytes::Bytes;
 use bytesize::ByteSize;
 use eyeball::SharedObservable;
+use http::header::CONTENT_LENGTH;
 use ruma::api::{
     client::error::{ErrorBody as ClientApiErrorBody, ErrorKind as ClientApiErrorKind},
     error::FromHttpResponseError,
@@ -166,14 +167,15 @@ pub(super) async fn send_request(
     let request = clone_request(request);
     let request = {
         let mut request = if send_progress.subscriber_count() != 0 {
-            send_progress.update(|p| p.total += request.body().len());
+            let content_length = request.body().len();
+            send_progress.update(|p| p.total += content_length);
 
             // Make sure any concurrent futures in the same task get a chance
             // to also add to the progress total before the first chunks are
             // pulled out of the body stream.
             tokio::task::yield_now().await;
 
-            reqwest::Request::try_from(request.map(|body| {
+            let mut req = reqwest::Request::try_from(request.map(|body| {
                 let chunks = stream::iter(BytesChunks::new(body, 8192).map(
                     move |chunk| -> Result<_, Infallible> {
                         send_progress.update(|p| p.current += chunk.len());
@@ -181,7 +183,14 @@ pub(super) async fn send_request(
                     },
                 ));
                 reqwest::Body::wrap_stream(chunks)
-            }))?
+            }))?;
+
+            // When streaming the request, reqwest / hyper doesn't know how
+            // large the body is, so it doesn't set the content-length header
+            // (required by some servers). Set it manually.
+            req.headers_mut().insert(CONTENT_LENGTH, content_length.into());
+
+            req
         } else {
             reqwest::Request::try_from(request)?
         };

